### ISSUE TITLE
Issue - Running kubeapps behind contour ingress controller

### ISSUE THREAD

**Issue by AllaeddineEL** [kubeapps/kubeapps#3716]
### Description:

I have an issue running kubeapps behind contour as an ingress controller, the issue start to show up from Kubeapps v2.4.0 (chart version 7.4.0)
It fails at the POST on `apis/kubeappsapis.core.packages.v1alpha1.PackagesService/GetInstalledPackageSummaries` with 404

![Bildschirmfoto 2021-11-03 um 09 19 28](https://user-images.githubusercontent.com/17249893/140028002-ead50dc0-82c9-4c9e-9f6f-f251a2ce4617.png)

#### Logs from kubeapp/nginx pod
```
100.96.3.92 - - [03/Nov/2021:08:30:37 +0000] "POST /apis/kubeappsapis.core.packages.v1alpha1.PackagesService/GetInstalledPackageSummaries HTTP/1.1" 404 32 "https://kubeapps.demo.mytanzusandbox.com/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/94.0.4606.61 Safari/537.36"
100.96.3.92 - - [03/Nov/2021:08:30:37 +0000] "GET /api/clusters/default/apis/operators.coreos.com/v1alpha1/namespaces/default/clusterserviceversions HTTP/1.1" 403 464 "https://kubeapps.demo.mytanzusandbox.com/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/94.0.4606.61 Safari/537.36"
```
with `kubectl port-forwarding` everything is working fine 
### Steps to reproduce the issue:

1. deploy contour as ingress controller
2. deploy kubeapps bitnami helm chart version 7.4.0

### Version of Helm, Kubeapps and Kubernetes:

- Output of `helm version`:

```
version.BuildInfo{Version:"v3.5.0", GitCommit:"32c22239423b3b4ba6706d450bd044baffdcf9e6", GitTreeState:"dirty", GoVersion:"go1.15.6"}
```

- Output of `helm list -n <kubeapps-namespace> <kubeapps-release-name>`:

```
NAME    	NAMESPACE	REVISION	UPDATED                             	STATUS  	CHART          	APP VERSION
kubeapps	kubeapps 	5       	2021-11-01 17:39:06.561647 +0100 CET	deployed	kubeapps-7.5.10	2.4.1
```

- Output of `kubectl version`:

```
Client Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.2+vmware.1", GitCommit:"54e7e68e30dd3f9f7bb4f814c9d112f54f0fb273", GitTreeState:"clean", BuildDate:"2021-06-28T22:26:48Z", GoVersion:"go1.16.5", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.2+vmware.1", GitCommit:"54e7e68e30dd3f9f7bb4f814c9d112f54f0fb273", GitTreeState:"clean", BuildDate:"2021-06-28T22:12:04Z", GoVersion:"go1.16.5", Compiler:"gc", Platform:"linux/amd64"}
```

**Comment by absoludity** [comment]
Thanks for the report.

> with `kubectl port-forwarding` everything is working fine

Hmm. This piece of info, together with the response headers shown above, would normally have me thinking that the requests aren't even making it to the kubeapps apis service (you can check by watching the log when using ingress verses when using port forwarding). But if those kubeapps/nginx logs are from requests using ingress (rather than port-forwarding), then it would appear the requests are getting through to the apis service. Can you please verify with the logs of the kubeapps apis service whether the request is getting there (compare with and without port-forwarding).

Either way, it could be that envoy, which has [native support for grpc and grpc-web](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/other_protocols/grpc), thinks that it should be handling these grpc-web requests with its filter and is somehow modifying the grpc requests (or not forwarding them due to a lack of config, above will show). The envoy logs for the same requests would also be useful.

Thanks!

**Comment by absoludity** [comment]
Thanks Timo. So the key bit of info that we're after is to see if that request is getting from the frontend component through to the kubeapps APIs service. For example, if I check my own local setup, I see the request in the frontend nginx container:
```
127.0.0.1 - - [07/Dec/2021:19:09:35 +0000] "POST /apis/kubeappsapis.core.packages.v1alpha1.PackagesService/GetInstalledPackageSummaries HTTP/1.1" 200  291 "https://localhost/" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:94.0) Gecko/20100101 Firefox/94.0" "172.18.0.1, 10.244.0.14"
```

and a corresponding request in the kubeapps APIs component, at the same time (well, slightly later):

```
I1207 19:09:35.416467       1 packages.go:180] +core GetInstalledPackageSummaries (cluster="default", namespace="default")
I1207 19:09:35.416756       1 server.go:667] +helm GetInstalledPackageSummaries (cluster="default", namespace="default")
```

This shows me that the request that is proxied by the frontend is actually reaching the kubeapps-apis service.

What we suspect is happening when contour is configured as ingress with a certain configuration (don't know yet), is that it is intercepting and/or modifying the request, such that it's no longer recognised as a grpc-web request by the time it gets to the kubeapps-apis service (if it gets there).

Can you check if you can see a corresponding entry in the logs of the kubeapps apis pod?

I'll re-open as we get more info. Thanks!

**Comment by tsalm-pivotal** [comment]
There are no such log entries for the kubeapps APIs components.

**Comment by absoludity** [comment]
> There are no such log entries for the kubeapps APIs components.

OK, that's useful info. So given that you're using Contour for the ingress controller, AIUI that means you must be using envoy in your cluster as well (since contour is just a control plane for the envoy proxy). [Envoy has built-in support for proxying and translating GRPC requests](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/other_protocols/grpc#grpc-bridging) which we (Kubeapps) explicitly *don't* want to use. If you have a chance, can you please check both the envoy logs as well as the envoy configuration (especially the grpc bridge), to see if it's handling/modifying/filtering the requests.

If not, at some point we'll setup an environment to reproduce locally and verify.

Thanks for the input Timo!

**Comment by tsalm-pivotal** [comment]
This is the envoy configuration used in the pod. It is autogenerated, so there is no easy way to change it.
```
{
  "static_resources": {
    "clusters": [
      {
        "name": "contour",
        "alt_stat_name": "tanzu-system-ingress_contour_8001",
        "type": "STRICT_DNS",
        "connect_timeout": "5s",
        "load_assignment": {
          "cluster_name": "contour",
          "endpoints": [
            {
              "lb_endpoints": [
                {
                  "endpoint": {
                    "address": {
                      "socket_address": {
                        "address": "contour",
                        "port_value": 8001
                      }
                    }
                  }
                }
              ]
            }
          ]
        },
        "circuit_breakers": {
          "thresholds": [
            {
              "priority": "HIGH",
              "max_connections": 100000,
              "max_pending_requests": 100000,
              "max_requests": 60000000,
              "max_retries": 50
            },
            {
              "max_connections": 100000,
              "max_pending_requests": 100000,
              "max_requests": 60000000,
              "max_retries": 50
            }
          ]
        },
        "typed_extension_protocol_options": {
          "envoy.extensions.upstreams.http.v3.HttpProtocolOptions": {
            "@type": "type.googleapis.com/envoy.extensions.upstreams.http.v3.HttpProtocolOptions",
            "explicit_http_config": {
              "http2_protocol_options": {}
            }
          }
        },
        "transport_socket": {
          "name": "envoy.transport_sockets.tls",
          "typed_config": {
            "@type": "type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.UpstreamTlsContext",
            "common_tls_context": {
              "tls_certificate_sds_secret_configs": [
                {
                  "name": "contour_xds_tls_certificate",
                  "sds_config": {
                    "path": "/config/resources/sds/xds-tls-certificate.json",
                    "resource_api_version": "V3"
                  }
                }
              ],
              "validation_context_sds_secret_config": {
                "name": "contour_xds_tls_validation_context",
                "sds_config": {
                  "path": "/config/resources/sds/xds-validation-context.json",
                  "resource_api_version": "V3"
                }
              }
            }
          }
        },
        "upstream_connection_options": {
          "tcp_keepalive": {
            "keepalive_probes": 3,
            "keepalive_time": 30,
            "keepalive_interval": 5
          }
        }
      },
      {
        "name": "service-stats",
        "alt_stat_name": "tanzu-system-ingress_service-stats_9001",
        "type": "STATIC",
        "connect_timeout": "0.250s",
        "load_assignment": {
          "cluster_name": "service-stats",
          "endpoints": [
            {
              "lb_endpoints": [
                {
                  "endpoint": {
                    "address": {
                      "socket_address": {
                        "address": "127.0.0.1",
                        "port_value": 9001
                      }
                    }
                  }
                }
              ]
            }
          ]
        }
      }
    ]
  },
  "dynamic_resources": {
    "lds_config": {
      "api_config_source": {
        "api_type": "GRPC",
        "transport_api_version": "V3",
        "grpc_services": [
          {
            "envoy_grpc": {
              "cluster_name": "contour"
            }
          }
        ]
      },
      "resource_api_version": "V3"
    },
    "cds_config": {
      "api_config_source": {
        "api_type": "GRPC",
        "transport_api_version": "V3",
        "grpc_services": [
          {
            "envoy_grpc": {
              "cluster_name": "contour"
            }
          }
        ]
      },
      "resource_api_version": "V3"
    }
  },
  "admin": {
    "access_log_path": "/dev/null",
    "address": {
      "socket_address": {
        "address": "127.0.0.1",
        "port_value": 9001
      }
    }
  }
}
```

**Comment by tsalm-pivotal** [comment]
Envoy log:
```
[2021-12-08T07:43:38.119Z] "POST /apis/kubeappsapis.core.packages.v1alpha1.PackagesService/GetAvailablePackageSummaries HTTP/2" 404 - 48 0 10 9 "100.96.3.1" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/97.0.4692.36 Safari/537.36" "10c477f3-c44a-41c0-8f7e-49bd549b8c07" "kubeapps.emea.end2end.link" "100.96.7.237:8080"
```

**Comment by absoludity** [comment]
Thanks again Timo. So yes, it appears that the requests are not correctly proxied to the kubeapps-apis server, but it's not clear from the configuration why that is. We'll need to setup a local env with contour/envoy and debug it closely. I'll put this issue up for discussion in our next planning meeting.

**Comment by satya-dillikar** [comment]
Hello Timo, 

I am unable to reproduce this issue in the latest version. Maybe I am missing something here.


My kubeapps helm-chart

```
helm list --all-namespaces
NAME    	NAMESPACE	REVISION	UPDATED                             	STATUS  	CHART         	APP VERSION
kubeapps	kubeapps 	1       	2021-12-15 11:12:55.334271 -0800 PST	deployed	kubeapps-7.6.2	2.4.2

```

```
kubectl version
Client Version: version.Info{Major:"1", Minor:"22", GitVersion:"v1.22.3", GitCommit:"c92036820499fedefec0f847e2054d824aea6cd1", GitTreeState:"clean", BuildDate:"2021-10-27T18:34:20Z", GoVersion:"go1.16.10", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.1", GitCommit:"5e58841cce77d4bc13713ad2b91fa0d961e69192", GitTreeState:"clean", BuildDate:"2021-05-21T23:01:33Z", GoVersion:"go1.16.4", Compiler:"gc", Platform:"linux/amd64"}
```

Contour
```

kubectl get all -n projectcontour
NAME                           READY   STATUS    RESTARTS   AGE
pod/contour-76bb4ff6cc-swvzs   1/1     Running   0          71m
pod/contour-76bb4ff6cc-zgnzj   1/1     Running   0          71m
pod/envoy-dv5bh                2/2     Running   0          71m

NAME              TYPE           CLUSTER-IP      EXTERNAL-IP   PORT(S)                      AGE
service/contour   ClusterIP      10.96.223.16    <none>        8001/TCP                     71m
service/envoy     LoadBalancer   10.96.179.123   <pending>     80:30926/TCP,443:30087/TCP   71m

NAME                   DESIRED   CURRENT   READY   UP-TO-DATE   AVAILABLE   NODE SELECTOR   AGE
daemonset.apps/envoy   1         1         1       1            1           <none>          71m

NAME                      READY   UP-TO-DATE   AVAILABLE   AGE
deployment.apps/contour   2/2     2            2           71m

NAME                                 DESIRED   CURRENT   READY   AGE
replicaset.apps/contour-76bb4ff6cc   2         2         2       71m
```

Ingress

```
 kubectl get ingress -n kubeapps
NAME       CLASS     HOSTS           ADDRESS   PORTS   AGE
kubeapps   contour   dev.satya.com             80      47m
```

Ingress profile: kubeapps-ingress.yaml

```
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: kubeapps
  annotations:
    cert-manager.io/cluster-issuer: letsencrypt
    nginx.ingress.kubernetes.io/rewrite-target: /
    nginx.ingress.kubernetes.io/proxy-read-timeout: 600s
    nginx.ingress.kubernetes.io/proxy-buffer-size: 16k
    nginx.ingress.kubernetes.io/proxy-buffers: "8"
    nginx.ingress.kubernetes.io/client-body-buffer-size: 8m
spec:
  ingressClassName: contour
  rules:
  - host: dev.satya.com
    http:
      paths:
      - backend:
          service:
            name: kubeapps
            port:
              number: 80
        path: /
        pathType: Prefix
```

My screenshot:
![kubeapps-contour](https://user-images.githubusercontent.com/37601587/146258321-90828de6-dc11-451f-afb3-631bbff6051d.png)


@tsalm-pivotal,
In the “Response Headers” of the screenshot. I notice you have
"content-type: application/grpc-web+proto"
Please let me know how you are triggering the API call.

Please compare my ingress profile as mentioned above and let us know if that works for you. 

Thanks
Satya

**Comment by satya-dillikar** [comment]
A quick update:  I am able to reproduce this with an AWS EKS cluster.  Not sure why it was not seen in the 'kind' cluster.

**Comment by migmartri** [comment]
Sorry I closed it by mistake!

**Comment by satya-dillikar** [comment]
@AllaeddineEL, @tsalm-pivotal, In an effort to narrow down the issue, I have replaced the 'kubeapps' app with other servers: 
(1) http-echo server & also  
(2) [grpc echo server](https://github.com/mhausenblas/yages) app for sample testing.

To my surprise, project contour didn't work with either of them even after adding **the annotations to the service.**

See my discussion on Kubernetes workspace slack#channel contour  [here](https://kubernetes.slack.com/team/U02JWUCKSH3)

I replaced project-contour with an **ingress-Nginx controller** and noticed that the http-echo server works as expected.

As we see, both http-echo and kubeapps are not working with project contour only.

So, I am guessing something is missing in service annotations which are very specific to project contour.

if anyone of you is available to debug, I can share my setup.

**Comment by satya-dillikar** [comment]
Service with annotations:

kind: Service
apiVersion: v1
metadata:
  name: banana-service
  namespace: nginx-sample-traffic
  annotations:
    projectcontour.io/upstream-protocol.h2c: "5678"
spec:
  selector:
    app: banana
  ports:
    - port: 5678 # Default port for image

---
kind: Service
apiVersion: v1
metadata:
  name: apple-service
  namespace: nginx-sample-traffic
  annotations:
    projectcontour.io/upstream-protocol.h2c: "5678"
spec:
  selector:
    app: apple
  ports:
    - port: 5678 # Default port for image

**Comment by satya-dillikar** [comment]
Ingress with contour:

apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: example-contour
  namespace: nginx-sample-traffic
spec:
  ingressClassName: contour
  rules:
  - host: a06589614a2b743688ef6b8cdbfc6912-751344421.us-west-2.elb.amazonaws.com
    http:
      paths:
      - backend:
          service:
            name: apple-service
            port:
              number: 5678
        path: /apple
        pathType: Prefix
      - backend:
          service:
            name: banana-service
            port:
              number: 5678
        path: /banana
        pathType: Prefix

**Comment by sunjayBhatia** [comment]
Hey there, Contour maintainer here

This issue was forwarded to me and messing around with it for a bit, everything seems to work when I completely remove the grpc-web filter (did so in a local build)

lines up with this comment `Envoy has built-in support for proxying and translating GRPC requests which we (Kubeapps) explicitly don't want to use.` which I just noticed after a more detailed read of this thread

Contour currently *always* adds the grpc-web filter to the Envoy filter chains for HTTP and HTTPS virtualhosts, so we would have to implement an optional toggle to remove it (or possibly use some of the newer Envoy features to skip a filter based on request attributes, though that is still alpha)

**Comment by satya-dillikar** [comment]
> From a very very cursory glance at some of the `kubeapps` code, it seems there is already some multiplexing of http/gRPC/gRPC-web going on? Any reason to not support gRPC and gPRC-Web for the endpoints in question?

Thanks Sunjay for looking into this.  kubeapps does support grpc and grcp-web.  You can try this on any grpc server  (not necessarily kubeapps).

**Comment by absoludity** [comment]
Yes, thanks @sunjayBhatia . Just to expand on what Satya mentioned, backend api server itself can be interacted with via `grpcurl` (for straight gPRC), `gRPC-Web` (using the Kubeapps browser client), or `http`, *but*, since the Kubeapps browser frontend only uses `gRPC-Web`, it is normally no issue that the endpoints in question are reverse-proxied via nginx (with oauth2-proxy as all our backend services are).

I have not been actively investigating this issue yet, but the [data from Timo above](https://github.com/kubeapps/kubeapps/issues/3716#issuecomment-988573807) appears to show that contour returns a 421 (misdirected request) before returning a 404 to the client. I wonder if contour is probing the endpoint and finding that it doesn't satisfy a grpc request, since it is reverse-proxied via nginx.

Actually, regardless, I wonder if it is simply that contour is doing the `gRPC-Web` -> `gRPC` translation in front of our nginx reverse proxy, which cannot work (currently).

**Comment by sunjayBhatia** [comment]
> the data from Timo above appears to show that contour returns a 421 (misdirected request) before returning a 404 to the client

I did not see that when I reproduced it, I also did not see any HTTP/2 requests arrive at Envoy as shown in that log snippet, only HTTP/1.1

**Comment by sunjayBhatia** [comment]
Envoy debug logs:

Request from browser (Envoy translates to HTTP/2 pseudo headers etc. internally for processing/logging)
```
[2022-01-24 16:54:17.924][27][debug][http] [source/common/http/conn_manager_impl.cc:867] [C61371][S3569982871577445907] request headers complete (end_stream=false):
':authority', 'example.com:9080'
':path', '/apis/kubeappsapis.core.packages.v1alpha1.PackagesService/GetInstalledPackageSummaries'
':method', 'POST'
'connection', 'keep-alive'
'content-length', '30'
'user-agent', 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/96.0.4664.55 Safari/537.36'
'x-grpc-web', '1'
'authorization', 'Bearer xxx'
'content-type', 'application/grpc-web+proto'
'accept', '*/*'
'origin', 'http://example.com:9080'
'referer', 'http://example.com:9080/'
'accept-encoding', 'gzip, deflate'
'accept-language', 'en-US,en;q=0.9'
```

```
[2022-01-24 16:54:17.925][27][debug][router] [source/common/router/router.cc:457] [C61371][S3569982871577445907] cluster 'kubeapps/kubeapps/80/da39a3ee5e' match for URL '/apis/kubeappsapis.core.packages.v1alpha1.PackagesService/GetInstall
edPackageSummaries'
```

Translated gPRC request to kubeapps:
```
[2022-01-24 16:54:17.925][27][debug][router] [source/common/router/router.cc:673] [C61371][S3569982871577445907] router decoding headers:
':authority', 'example.com'
':path', '/apis/kubeappsapis.core.packages.v1alpha1.PackagesService/GetInstalledPackageSummaries'
':method', 'POST'
':scheme', 'http'
'user-agent', 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/96.0.4664.55 Safari/537.36'
'x-grpc-web', '1'
'authorization', 'Bearer xxx'
'content-type', 'application/grpc'
'accept', '*/*'
'origin', 'http://example.com:9080'
'referer', 'http://example.com:9080/'
'accept-encoding', 'gzip, deflate'
'accept-language', 'en-US,en;q=0.9'
'x-forwarded-for', '172.23.0.1'
'x-forwarded-proto', 'http'
'x-envoy-internal', 'true'
'x-request-id', 'd446a9ed-915e-4f0c-b060-1c9868a058b7'
'te', 'trailers'
'grpc-accept-encoding', 'identity'
'x-envoy-expected-rq-timeout-ms', '15000'
'x-request-start', 't=1643043257.924'
```

Response
```
[2022-01-24 16:54:17.926][27][debug][http] [source/common/http/conn_manager_impl.cc:1467] [C61371][S3569982871577445907] encoding headers via codec (end_stream=false):
':status', '404'
'server', 'envoy'
'date', 'Mon, 24 Jan 2022 16:54:17 GMT'
'content-type', 'application/grpc-web+proto'
'content-length', '0'
'x-envoy-upstream-service-time', '0'
'grpc-message', '{"code":5,"message":"Not Found"}'
'grpc-status', '12'
```

Envoy access log
```
[2022-01-24T16:54:17.924Z] "POST /apis/kubeappsapis.core.packages.v1alpha1.PackagesService/GetInstalledPackageSummaries HTTP/1.1" 404 - 30 0 1 0 "172.23.0.1" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like
 Gecko) Chrome/96.0.4664.55 Safari/537.36" "d446a9ed-915e-4f0c-b060-1c9868a058b7" "example.com" "10.244.1.68:8080"
```

**Comment by absoludity** [comment]
> > I wonder if it is simply that contour is doing the gRPC-Web -> gRPC translation in front of our nginx reverse proxy, which cannot work
> 
> yep, thats what the Envoy gRPC-Web filter is doing so requests to the kubeapps frontend are translated to gRPC when Contour is the ingress controller since we make an assumption that anyone using gRPC-Web will want that

Thanks for the detail @sunjayBhatia . This is then the issue. Our nginx reverse proxy is currently proxying grpc-web requests only (we're not using `grpc_pass` at all) since this is what our frontend uses.

So if the contour grpc-filter was able to be switched off, this would work with Kubeapps OOTB. We may also be able to use the `grpc_proxy` nginx option to additionally proxy plain grpc requests through as well, which would enable Kubeapps to work with contour OOTB.

**Comment by sunjayBhatia** [comment]
Yep, when I recompiled contour and removed the gRPC-Web filter kubeapps seemed to work fine, we have an issue at Contour to look into making that filter optional, though it’s not quite straightforward: https://github.com/projectcontour/contour/issues/4290

**Comment by absoludity** [comment]
> We'll need to setup a local env with contour/envoy and debug it closely.

I finally got a chance to look at this more closely, setting up Contour/Envoy in my local dev environment, to see the behaviour @sunjayBhatia mentioned above. Additionally, I updated the kubeapps-apis service with a debug log for every request to see that the request does indeed get through to our grpc api server, but as an http 1.0 request (added formatting):

```
E0310 03:06:38.364903       1 server.go:127] Request was: &{
Method:POST 
URL:/kubeappsapis.plugins.resources.v1alpha1.ResourcesService/CheckNamespaceExists 
Proto:HTTP/1.0 
ProtoMajor:1 
ProtoMinor:0 
Header:map[
  Accept:[*/*] 
  Accept-Encoding:[gzip, deflate, br] 
  Accept-Language:[en-US,en;q=0.5] 
  Authorization:[Bearer ...] 
  Cache-Control:[no-cache] 
  Connection:[close] 
  Content-Length:[25] 
  Content-Type:[application/grpc] 
  Dnt:[1] 
  Grpc-Accept-Encoding:[identity] 
  Origin:[https://localhost] 
  Pragma:[no-cache] 
  Referer:[https://localhost/] 
  Sec-Fetch-Dest:[empty] 
  Sec-Fetch-Mode:[cors] 
  Sec-Fetch-Site:[same-origin] 
  User-Agent:[Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:97.0) Gecko/20100101 Firefox/97.0] 
  X-Envoy-Expected-Rq-Timeout-Ms:[15000] 
  X-Envoy-Internal:[true] 
  X-Forwarded-For:[172.18.0.1] 
  X-Forwarded-Proto:[https] 
  X-Grpc-Web:[1] 
  X-Request-Id:[0d27ac83-1958-4383-8fc6-4ec2da19e208] 
  X-Request-Start:[t=1646881598.364]
] 
Body:0xc00074e9c0 
GetBody:<nil> 
ContentLength:25 
TransferEncoding:[] 
Close:true 
Host:kubeapps-internal-kubeappsapis:8080 
Form:map[] 
PostForm:map[] 
MultipartForm:<nil> 
Trailer:map[] 
RemoteAddr:10.244.0.35:49020 
RequestURI:/kubeappsapis.plugins.resources.v1alpha1.ResourcesService/CheckNamespaceExists 
TLS:<nil> 
Cancel:<nil> 
Response:<nil> 
ctx:0xc00074ea00}
```

I suspect the reason it arrives as an http 1.0 request  is because our nginx frontend is http1.1 only (not 2), so when envoy realises it can't send the gRPC request using http/2 it defaults to 1.0? But either way, it's not currently recognised as a valid gRPC request by our mux, since it's not http/2, nor is it worth us updating to recognise an http 1.0 grpc request (we need trailers and streaming etc.)

So if we were to solve this on Kubeapps' side (in addition to an option on contour's side with https://github.com/projectcontour/contour/issues/4290), why not add grpc proxying to our nginx config? To do so requires supporting http/2 and if we enable http2 for our nginx (without tls), then it can only be reached with --http2-prior-knowledge

```
$ curl --http2 localhost:8080 
curl: (52) Empty reply from server

$ curl localhost:8080 
curl: (52) Empty reply from server

curl --http2-prior-knowledge localhost:8080

<!doctype html><html lang="en"><head><link href="./static/css/2.abae2cf5.chunk.css" rel="stylesheet"><link href="./static/css/main.415df5f...
...

```

It is a [known issue for nginx](https://trac.nginx.org/nginx/ticket/816) that nginx cannot serve both http/1.1 and http/2 on the same port without TLS (but with TLS it can).

So one option is to update our nginx to use tls, so external requests have tls terminated at ingress (as it does now), but then the forwarded request between the ingress and our nginx frontend would also be over tls. This is possible, but would require cert creation and configuration for our internal services, which we want to do for other reasons anyway eventually (see #2268), I'm just not sure how much would be involved to do so (need to dig more).

The other option may be to use the [matching API](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/advanced/matching/matching_api) option of envoy (which @sunjayBhatia mentioned on https://github.com/projectcontour/contour/issues/4290 ) to configure convoy so that it doesn't do the grpcweb->grpc translation for our requests (I want to test this locally to confirm whether it works/fixes the issue).
1. Update our nginx config to allow http2 in addition to http1.1.

**Comment by absoludity** [comment]
Update: the matching API option isn't yet an option yet (see [slack conversation](https://kubernetes.slack.com/archives/C8XRH2R4J/p1646958856114189)). I had exhausted my ideas today and was about to give up, when I realised we don't even need to send the grpc-web requests through our nginx frontend: we can instead add extra paths to the ingress to route the `/apis/*` requests directly from the contour ingress through to the kubeapps-apis backend.

But for some reason still unknown, although the requests are routed directly (and are translated and arriving as application/grpc), the requests are still not recognised by our service as gRPC requests, rather being handled by the http gRPC+web handler (which still returns not found). I can talk directly via grpcurl without issue, so something is different between the envoy-translated grpc requests and what grpcurl sends.

Next step is to dump/debug what's actually arriving at the kubeapps-apis server and see what the difference is.

**Comment by absoludity** [comment]
Just re-opening and detailing the current status after doing some in-depth investigation.

TL;DR: It is possible to run Kubeapps with the current contour (ie. without https://github.com/projectcontour/contour/issues/4290 ) when using the demo-only, insecure, token authentication. I am not able to get Kubeapps working with Contour with OIDC. This may suite some users, but token authentication is something we recommend against using for anything other than demos.

## Kubeapps with Contour and token authentication

To ensure that the gRPC+web requests from the browser which are translated by Contour to gRPC requests can be received at our kubeapps-apis backend as gRPC requests, we need to both:

- Ensure that contour uses http2 over TCP (h2c)
- Ensure that the `/apis/` prefix from the gRPC-web request is removed (TODO: we could update so as not to require this prefix, I assume, with nginx routing on the request type instead?)

It isn't possible to do the prefix stripping with Contour's Ingress, but is possible with Contour's `HTTPProxy` (which, from reading the docs, is a more feature rich Ingress without relying on the annotation mess for extended functionality). So after installing Kubeapps configured for token authentication, with an existing tls cert for the domain (`localhost` in the example below), it turns out that all is required is the following `HTTPProxy` resource:

```
cat <<EOF | kubectl apply -f - 
apiVersion: projectcontour.io/v1
kind: HTTPProxy
metadata:
  name: kubeapps-grpc
  namespace: kubeapps
spec:
  virtualhost:
    fqdn: localhost
    tls:
      secretName: localhost-tls
  routes:
    - conditions:
      - prefix: /apis/
      pathRewritePolicy:
        replacePrefix:
        - replacement: /
      services:
      - name: kubeapps-internal-kubeappsapis
        port: 8080
        protocol: h2c
    - services:
      - name: kubeapps
        port: 80
EOF
```

This ensures that any requests to our kubeapps-apis service (ie. prefix `/apis/`) are not sent via the nginx frontend, but rather are routed directly by the Contour `HTTPProxy` through to the service using h2c and stripping the prefix in the process.

With this, Kubeapps appears to function perfectly fine with gRPC+web requests translated by contour.

## Kubeapps with Contour and OIDC authentication

The default configuration for Kubeapps with OIDC authentication has *every* request going through the `oauth2-proxy` service on the Kubeapps frontend, which replaces the authentication cookie with an Authorization header as the request passes through.

After a few different attempts, it appears that `oauth2-proxy` does not speak h2c:
```
< grpc-message: upstream connect error or disconnect/reset before headers. reset reason: protocol error
< grpc-status: 14
```

If it did, we may have been able to add an [extra upstream to our oauth2-proxy config](https://oauth2-proxy.github.io/oauth2-proxy/docs/configuration/overview#upstreams-configuration), so that requests for `/apis/` are sent directly from oauth2-proxy through to the kubeapps-apis backend. But as it is, the only way I can see us getting the current Contour with Kubeapps OIDC auth is if the contour `HTTPProxy` could switch the cookie for the `Authorization` header (by querying oauth2-proxy?) but I don't see that as being possible at either end.

So unfortunately, for Kubeapps with OIDC, we need to wait for https://github.com/projectcontour/contour/issues/4290 .

**Comment by cmoulliard** [comment]
Remark: The `secretName: localhost-tls` should be removed from the `HTTPProxy` manifest defined for "Kubeapps with Contour and token authentication" as that is failing.

**Comment by absoludity** [comment]
> Remark: The `secretName: localhost-tls` should be removed from the `HTTPProxy` manifest defined for "Kubeapps with Contour and token authentication" as that is failing.

Thanks Charles. I think it depends on your setup. In my local dev environment, I'm using the contour HTTPProxy/Ingress for TLS termination, which is why it's there, but yes, it won't be required (and will cause an error) if you're not doing so.
